pallets/parachain-system/src/validate_block/implementation.rs

```diff
@@ -77,39 +77,41 @@ pub fn validate_block<
 	PSC: crate::Config,
 	CI: crate::CheckInherents<B>,
 >(
-	params: ValidationParams,
+	MemoryOptimizedValidationParams {
+		block_data,
+		parent_head,
+		relay_parent_number,
+		relay_parent_storage_root,
+	}: MemoryOptimizedValidationParams,
 ) -> ValidationResult
 where
 	B::Extrinsic: ExtrinsicCall,
 	<B::Extrinsic as Extrinsic>::Call: IsSubType<crate::Call<PSC>>,
 {
-	let block_data = ParachainBlockData::<B>::decode(&mut &params.block_data.0[..])
+	let block_data = codec::decode_from_bytes::<ParachainBlockData<B>>(block_data)
 		.expect("Invalid parachain block data");
 
-	let parent_head =
-		B::Header::decode(&mut &params.parent_head.0[..]).expect("Invalid parent head");
-
-	// We decoded the block data and don't need it anymore. So, we release it to free the memory.
-	mem::drop(params.block_data);
+	let parent_header =
+		codec::decode_from_bytes::<B::Header>(parent_head.clone()).expect("Invalid parent head");
 
 	let (header, extrinsics, storage_proof) = block_data.deconstruct();
 
 	let head_data = HeadData(header.encode());
 
 	let block = B::new(header, extrinsics);
-	assert!(parent_head.hash() == *block.header().parent_hash(), "Invalid parent hash");
+	assert!(parent_header.hash() == *block.header().parent_hash(), "Invalid parent hash");
 
 	let inherent_data = extract_parachain_inherent_data(&block);
 
 	validate_validation_data(
 		&inherent_data.validation_data,
-		params.relay_parent_number,
-		params.relay_parent_storage_root,
-		params.parent_head,
+		relay_parent_number,
+		relay_parent_storage_root,
+		parent_head,
```

Comment:
As I've said in my previous comment, this could be dropped earlier if we'd compare the hashes instead of doing a bytewise comparison of `parent_head` to `validation_data.parent_head` (which, assuming the hash is a cryptographic hash, would be safe).

I'm not sure if it makes a difference (it only would if there could have conceivably be an allocation before this point which could be put in the same bucket as the full deallocated blob with the parameters, which I'd guess there isn't, so it most likely won't make a difference) and I'm fine with it as-is, but just like to make sure that this is an explicit decision and you didn't accidentally skip over my comment. (:

Reply:
> The problem being that hash could be implemented in any way (will probably not happen, but possible as it is a trait method).

Well, we could always just calculate a concrete hash that we know is secure. :P

But yeah, it's fine as-is, and might not be worth it. This is just my OCD pointing out the opportunity. :P